### PR TITLE
refactor: simplify the core and split the monolithic modules into packages

### PR DESCRIPTION
**Pull request by vboussot**
One squashed commit from the `refactor/deep-simplify` work.

**Simplification**: dead code, legacy paths and debug blocks removed; each duplicated piece of logic has one owner (`konfai.utils.clock`, `konfai.utils.vram`, `DataLog.parse`, `Network.bind`, `ImpactFeatureModel` behind the IMPACT criteria, `RemoteServer.get_json`, one cgroup/proc parse, rank-generic affine matrices, one `OutputDataset`; `OutSameAsGroupDataset` stays as an alias because published `Prediction.yml` name it).

**Packages**: the nine monolithic modules become packages, one module per family, bodies moved verbatim: `data/transform`, `utils/dataset` (one module per storage backend, `Dataset.<Backend>` kept), `predictor`, `data/data_manager`, `network/network`, `metric/measure`, `data/augmentation`, `utils/runtime`, `data/patching`. Every package re-exports its names: bare names in configs, published classpaths and `Dataset.<Backend>` resolve as before. `konfai-mcp` names a component by the package that re-exports it.

**Tests and docs**: shared geometry/manager builders, tests above eight seconds marked `slow`, the declared per-rank budget reset around every unit test (the old xdist flake is gone), tests patch the module that binds a name; comments keep the constraint and the measured figure; each package is documented by its modules.

**Verified**: core suite (fast, slow, integration), konfai-mcp, konfai-apps, impact_reg, studio; Sphinx at its baseline warnings. Computed values unchanged.

**Note for the merge**: the uncommitted work on `konfai/data/patching.py` in the shared checkout (the `_Unresolved` sentinel) will need porting to `data/patching/budget.py` / `manager.py`.

## Summary by CodeRabbit

* **New Features**
  * Added expanded transformation, augmentation, patching, dataset, streaming, prediction, metrics, and distributed-processing capabilities.
  * Added model ensembles, nested inference, additional dataset formats, streamed writes, and public component discovery.

* **Bug Fixes**
  * Improved prediction outputs, memory-budget recovery, statistics, resampling, and remote status requests.

* **Documentation**
  * Updated guides, examples, CLI references, and component documentation for the reorganized package structure.

* **Removed**
  * Removed DDPM, debug-only network blocks, and deprecated resume option formats.

### CODE REVIEW

File: konfai-mcp/konfai_mcp/server_support.py [hunk 1]
```diff
@@ -813,7 +810,7 @@ def _imported_object_extras(obj: Any) -> dict[str, Any]:
     extras: dict[str, Any] = {"bases": [], "forward": None, "konfai_base": None, "integration_hint": None}
     if not inspect.isclass(obj):
         return extras
-    mro = [f"{cls.__module__}.{cls.__qualname__}" for cls in obj.__mro__ if cls is not object]
+    mro = [f"{public_module(cls)}.{cls.__qualname__}" for cls in obj.__mro__ if cls is not object]
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai-mcp/konfai_mcp/server_support.py:818]: Fixed differently: the component base is now detected by issubclass against the imported base classes, and public_module only collapses to a package that re-exports every public name of the defining module, so Transform stays konfai.data.transform.Transform and konfai.data.reduction keeps its name.

File: konfai/api.py [hunk 2]
```diff
@@ -145,16 +146,24 @@ def _yaml_safe(value: object, where: str) -> object:
...
+        (m for m in default_modules if cls.__module__ == m or cls.__module__.startswith(m + ".")), cls.__module__
+    )
+
+
 def _classpath(stage: object, default_modules: tuple[str, ...]) -> str:
     """The name the config tree references ``stage`` by: bare for a shipped class, qualified else."""
     cls = type(stage)
-    return cls.__name__ if cls.__module__ in default_modules else f"{cls.__module__}:{cls.__name__}"
+    module = _public_module(cls, default_modules)
+    return cls.__name__ if module in default_modules else f"{module}:{cls.__name__}"
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/api.py:160]: Deliberate: for the names both packages define (Mask, Flip, Permute, Foreign), the bare spelling is the contract. TransformLoader resolves a bare name against transforms before the Expand marker and against augmentations after it (AGENTS.md section 4), so the round-trip preserves the stage type by position; a module-qualified spelling is emitted only for classes outside the shipped packages.

File: konfai/network/network/base.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            on_cuda
+            and not kwargs.get("differentiable", False)
+            and all(parameter.is_floating_point() for parameter in parameters)
+        )
+        if batchable and kwargs.get("fused") is None and kwargs.get("foreach") is None:
+            for name in ("fused", "foreach"):
+                if name in signature.parameters:
+                    kwargs[name] = True
+                    break
+        return optimizer_class(params, **kwargs)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/network/network/base.py:81]: Out of scope for this refactor: batched_step is moved verbatim and changing fused/foreach selection changes the optimizer's numerics. CPU training with the current selection runs in the integration suite (auto-patch training) without the described RuntimeError on the supported torch.

File: konfai/utils/config.py [hunk 4]
```diff
@@ -316,22 +315,12 @@ def __enter__(self):
...
-        if os.environ["KONFAI_CONFIG_MODE"] == "remove":
-            if os.path.exists(config_file()):
-                os.remove(config_file())
-            return
         # Only the visited subtree is folded back; the merge preserves the rest of the tree untouched.
-        subtree = self.create_dictionary(self.config, self.keys, len(self.keys) - 1)
+        subtree = self.config
+        for key in reversed(self.keys):
+            if key not in subtree:
+                subtree = {key: subtree}
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/config.py]: Fixed: the wrap is unconditional now; __enter__ always descends to the leaf, so the guard could only misfire on a leaf key named like a path component.

File: tests/unit/test_golden_run.py [hunk 5]
```diff
@@ -158,7 +158,7 @@ def _predict_config(dataset: Path) -> dict:
             "outputs_dataset": {
                 "Head:Softmax": {
                     "OutputDataset": {
-                        "name_class": "OutSameAsGroupDataset",
+                        "name_class": "OutputDataset",
```
[the remark this answers is not in the document]
Comment (vboussot) [reply tests/unit/test_golden_run.py:161]: Covered: test_the_published_sink_name_resolves_to_the_output_dataset (tests/unit/test_predictor.py) pins the alias through the same get_module resolution OutputDatasetLoader uses, which is what published configs exercise.

File: konfai/data/patching/blend.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+    def _window_1d(self, size: int, overlap: int) -> torch.Tensor:
+        window = torch.ones(size)
+        # sin**2 ramp over the overlap; the neighbouring patch's cos**2 ramp is its complement, so the two
+        # sum to exactly one across the overlap. The +0.5 phase keeps the very edge > 0 so a single-patch
+        # border, whose share is then the whole of the weight there, recovers the raw value.
+        ramp = torch.sin((torch.arange(overlap, dtype=torch.float32) + 0.5) / overlap * (torch.pi / 2)) ** 2
+        window[:overlap] = ramp
+        window[size - overlap :] = ramp.flip(0)
+        return window
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/data/patching/blend.py:169]: Fixed: the integer branch of blend_overlap now applies the same guard as resolve_overlap (overlap must be smaller than the patch size on every tiled axis), so an oversized integer overlap is refused at configuration instead of corrupting the taper.

File: konfai/data/patching/blend.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+    The blend has no volume extent: every axis whose patch is longer than one voxel is treated as tiled
+    (the untiled-axis-0 rule is already applied by the slicing plan), so an int is the same voxel
+    overlap on every such axis.
+    """
+    if isinstance(overlap, int):
+        for size in patch_size:
+            if size > 1 and overlap >= size:
+                raise ConfigError(f"overlap: {overlap} voxels must be smaller than the patch size {size}.")
+        return [overlap if size > 1 else 0 for size in patch_size]
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/data/patching/blend.py:147]: Fixed in 64a050e5: a negative integer overlap is refused with the same ConfigError wording as resolve_overlap.

File: konfai/utils/ome_zarr.py [hunk 8]
```diff
@@ -624,7 +640,18 @@ def _read_level_window(store_path: str, level: int, image: Any, index: tuple) ->
...
+    return _lazy_window(image.data, index)
+
+
+def _lazy_window(data: Any, index: tuple) -> np.ndarray:
+    """The plain lazy read. A lazy array that refuses a stepped slice (ngff-zarr >= 0.44 wraps the
+    level in an adapter that takes unit steps only) is read over the unit-step span and stepped here."""
+    try:
+        return np.asarray(data[index])
+    except NotImplementedError:
+        span = tuple(slice(item.start, item.stop) if isinstance(item, slice) else item for item in index)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/ome_zarr.py]: Fixed in 64a050e5: the fallback builds the ascending unit-step span for a negative-step slice and reads it backwards from its own end, which lands on the indices the original slice named.

File: konfai/utils/ome_zarr.py [hunk 9]
```diff
@@ -934,7 +959,7 @@ def create_ome_zarr_store(
         _type_component_axis(multiscales, _DISPLACEMENT_AXIS_TYPE)
         version = _RFC5_VERSION
     # version is explicit because to_ngff_zarr defaults to 0.5, which zarr-python 2 cannot write.
-    ngff_zarr.to_ngff_zarr(str(store_path), multiscales, overwrite=True, version=version)
+    _write_skeleton(store_path, multiscales, version)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/ome_zarr.py:1006]: Fixed in 64a050e5: append_ome_zarr_levels refuses a remote root with a structured DatasetManagerError before any Path() conversion; level derivation is a local operation, and the message says to write locally and upload.

File: apps/impact_reg/impact_reg_konfai/models/fireants.py [hunk 10]
```diff
@@ -414,20 +410,29 @@ def _stats(tensor: torch.Tensor) -> dict:
...
         # FireANTs' masked mode carries the mask as one extra trailing channel on BOTH images (see
         # ``apply_mask_to_image``). The feature model wants the image alone, so split the mask off and
         # hand it to KonfAI's masked feature loss (nearest-resampled onto every feature layer): the
         # metric is then evaluated inside the fixed mask, which is what an elastix/ITK mask means too.
         mask: torch.Tensor | None = None
-        if moved.shape[1] == self.in_channels + 1 and fixed.shape[1] == self.in_channels + 1:
+        in_channels = self.model.in_channels
+        if moved.shape[1] == in_channels + 1 and fixed.shape[1] == in_channels + 1:
             mask = (fixed[:, -1:] > 0.5).to(torch.uint8)
             moved, fixed = moved[:, :-1], fixed[:, :-1]
```
[the remark this answers is not in the document]
Comment (vboussot) [reply apps/impact_reg/impact_reg_konfai/models/fireants.py]: Fixed in 0081312. The masked state is now an explicit contract instead of a width inference: the engine passes its own `masked` decision into `ImpactFeatureLoss(specs, masked=...)`, the loss splits the trailing mask channel off once in `forward` and hands it to every core, and `_ImpactCore.forward` takes `mask` as a parameter. `in_channels` no longer participates in mask detection, so a 1-channel image with a 3-channel feature extractor is handled correctly.

File: konfai/data/transform/labels.py [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+        del source_spatial_shape, name
+        _forget_model_channel_counts(cache_attribute)
+
+    def __call__(self, name: str, tensor: torch.Tensor, cache_attribute: Attribute) -> torch.Tensor:
+        if "number_of_channels_per_model" in cache_attribute:
+            number_of_channels = cache_attribute.pop_tensor("number_of_channels_per_model")
+            result = tensor[0]
+            for i, t in enumerate(tensor[1:]):
+                t[t != 0] += int(number_of_channels[i]) - 1
+                result += t
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/data/transform/labels.py:239]: Known and deliberate, out of scope for this refactor (values unchanged everywhere). The non-cumulative offset in `Sum`'s multi-model branch dates from 1.4.0; published configs since then depend on its exact output, so the branch is frozen on purpose. The correct merge path is the `MergeLabels` transform, which accumulates the offset and clones; the published HF configs are being migrated from `Sum` to `MergeLabels`.
